core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/wordCloud/WordCloudOpFinalExec.java

```diff
@@ -38,42 +46,79 @@ public void close() {
         termFreqMap = null;
     }
 
-    @Override
-    public String getParam(String query) {
-        return null;
+    public List<Tuple> normalizeWordCloudTuples() {
+        double minValue = Double.MAX_VALUE;
+        double maxValue = Double.MIN_VALUE;
+
+        for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
+            int frequency = e.getValue();
+            minValue = Math.min(minValue, frequency);
+            maxValue = Math.max(maxValue, frequency);
+        }
+        // normalize the font size for wordcloud js
+        // https://github.com/timdream/wordcloud2.js/issues/53
+        List<Tuple> termFreqTuples = new ArrayList<>();
+        for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
+            termFreqTuples.add(Tuple.newBuilder().add(
+                    resultSchema,
+                    Arrays.asList(
+                            e.getKey(),
+                            (int) ((e.getValue() - minValue) / (maxValue - minValue) *
+                                    (MAX_FONT_SIZE - MIN_FONT_SIZE) + MIN_FONT_SIZE)
+                    )
+            ).build());
+        }
+        return termFreqTuples;
+    }
+
+    public List<Tuple> calculateResults(List<Tuple> normalizedWordCloudTuples) {
+        List<Tuple> retractions = new ArrayList<>(prevWordCloudTuples);
+        List<Tuple> insertions = new ArrayList<>(normalizedWordCloudTuples);
+
+        retractions.removeAll(normalizedWordCloudTuples);
+        insertions.removeAll(prevWordCloudTuples);
+
+        List<Tuple> results = new ArrayList<>();
+        retractions.forEach(tuple -> results.add(ProgressiveUtils.addRetractionFlag(tuple)));
+        insertions.forEach(tuple -> results.add(ProgressiveUtils.addInsertionFlag(tuple)));
+        return results;
```

Comment:
In my understanding, there are multiple copies going on here. 
1. adding a flag to a `Tuple` will make a new copy of it. 
2. retraction and insertion filter each other would require making a copy then filter. 
3. adding the final tuple into the `results` list would also make a copy.

Is there any concern here?

instead, if we just output the original tuple list, along with header information that refers to which tuple is an insertion and which tuple is a retraction, would it be better?

Reply:
This is out of scope of this PR. But the copy will be on the object level, not the underlying data.